### PR TITLE
Make demo script more portable and CI friendly

### PR DESCRIPTION
**Pull request by chrisburr**
By default macOS comes with an ancient version of bash (as zsh is the "preffered" shell). Normally I'd advocate for sticking to POSIX-sh for compatibility but the demo script is complex enough that it's nice to use actual bash and deal with the platform specificities.

### CODE REVIEW

File: run_demo.sh [hunk 1]
```diff
@@ -40,33 +42,96 @@ function check_hostname(){
...
+
+	# # Expand parametric values
+	# -[monadic-short-opts]?*|--[!=]*=*)
+	# 	case $1 in
+	# 		--*) tail=${1#*=}; head=${1%%=*} ;;
+	# 		*)   tail=${1#??}; head=${1%"$tail"} ;;
+	# 	esac
+	# 	shift
+	# 	set -- "$head" "$tail" "$@"
+	# 	continue ;;
```
Comment (aldbr) [inline run_demo.sh:71]: Is this expected?

File: run_demo.sh [hunk 2]
```diff
@@ -40,33 +42,96 @@ function check_hostname(){
...
-printf "%b Found package directories for: " ${UNICORN_EMOJI}
-echo "${pkg_names[@]}"
+if [ ${#pkg_names[@]} -gt 0 ]; then
+  printf "%b Found package directories for: %s\n" ${UNICORN_EMOJI} "${pkg_names[@]}"
+else
+  printf "%b No source directories were specified\n" ${UNICORN_EMOJI}
+fi
 
+# Try to find a suitable hostname/IP-address for the demo. This must be not
+# resolve to a loopback address as pods need to be able to commuinicate with
```
Comment (aldbr) [inline run_demo.sh:131]:
```suggestion
# resolve to a loopback address as pods need to be able to communicate with
```

File: run_demo.sh (hunk 1 above)
Comment (chrisburr) [reply run_demo.sh:71]: Yeah we currently don't have any `niladic` or `monadic` options but I think this is fiddley enough to write in portable bash that it's worth having commented out in preparation :)

File: run_demo.sh [hunk 3]
```diff
@@ -40,33 +42,96 @@ function check_hostname(){
...
-diracx_src_dir="${diracx_repo_dir}/src/diracx"
-if [[ ! -d "${diracx_src_dir}" ]]; then
-  printf "\U26A0\UFE0F Error: %s is not a clone of DiracX!" "${diracx_repo_dir}"
-  exit 1
-fi
+usage="${0##*/} [-h|--help] [--exit-when-done] [--] <diracx_src_dir> [other source directories]"
+
+# Parse command-line switches
+exit_when_done=0
+while [ -n "$1" ]; do case $1 in
```
[the remark this answers is not in the document]
Comment (chrisburr) [reply run_demo.sh:49]: Ah that's a bug as we should be able to run with no options 😄

File: run_demo.sh [hunk 4]
```diff
@@ -40,33 +42,96 @@ function check_hostname(){
...
+esac; shift
+done
+
+# Remaining arguments are positional parameters that are used to specify which
+# source directories to mount in the demo cluster
+declare -a pkg_dirs=()
+declare -a pkg_names=()
 for src_dir in "$@"; do
+  # shellcheck disable=SC2044
   for pkg_dir in $(find "$src_dir/src" -type f -mindepth 2 -maxdepth 2 -name '__init__.py'); do
```
Comment (aldbr) [inline run_demo.sh:102]: This is not related to this PR but should be easy to change:

```suggestion
  for pkg_dir in $(find "$src_dir/src" -mindepth 2 -maxdepth 2 -type f -name '__init__.py'); do
```

Adding `-type f` before `-min/maxdepth` generates the following warning:

```bash
find: warning: you have specified the global option -mindepth after the argument -type, but global options are not positional, i.e., -mindepth affects tests specified before it as well as those specified after it.  Please specify global options before other arguments.
find: warning: you have specified the global option -maxdepth after the argument -type, but global options are not positional, i.e., -maxdepth affects tests specified before it as well as those specified after it.  Please specify global options before other arguments.
```

**Review by aldbr — APPROVED** [review]

**Pull request merged**